### DISCUSSION TITLE
unmountInactiveRoutes / unmountOnBlur for Stack Navigation

### DISCUSSION THREAD

**Discussion by lcsjunior** [react-navigation/react-navigation#10987] [Q&A] [answered]
### Current behavior

1. Is there really no unmountInactiveRoutes / unmountOnBlur for Stack Navigation?
2. What's the way to do this, using useFocusEffect?
3. What do large projects that use React Navigation do?

Where I work, for example in CRUD, we reset all screen states, refs and hooks in useFocusEffect. Is this conventional?
The same thing goes when using [SWR](https://github.com/vercel/swr) or [nandorojo/swr-react-native](url) libraries.

Stack Navigation:
[https://snack.expo.dev/@lcsjuniorsi/rn-stack](url)
![stack](https://user-images.githubusercontent.com/8947621/200663710-4a13a7ca-8d22-4097-9915-50716e9dd095.gif)

Drawer Navigation "OK":
[https://snack.expo.dev/@lcsjuniorsi/rn-drawer](url)
![drawer](https://user-images.githubusercontent.com/8947621/200663746-428b1830-e110-47a9-8ef9-9d85644fe5d6.gif)

### Expected behavior

unmountInactiveRoutes / unmountOnBlur for Stack Navigation!

### Reproduction

https://snack.expo.dev/@lcsjuniorsi/rn-stack

### Platform

- [X] Android
- [X] iOS
- [X] Web
- [ ] Windows
- [ ] MacOS

### Packages

- [ ] @react-navigation/​bottom-tabs
- [X] @react-navigation/​drawer
- [ ] @react-navigation/​material-bottom-tabs
- [ ] @react-navigation/​material-top-tabs
- [X] @react-navigation/​stack
- [X] @react-navigation/​native-stack

### Environment

- [] I've removed the packages that I don't use

| package                                | version |
| -------------------------------------- | ------- |
| @react-navigation/native               | *
| @react-navigation/drawer               | 6.3.3
| @react-navigation/stack                | 6.3.3
| react-native-safe-area-context         | 4.3.1
| react-native-screens                   | ~3.15.0
| react-native-gesture-handler           | ~2.5.0
| react-native-reanimated                | ~2.9.1
| react-native                           | *
| expo                                   | *

**Comment by satya164** [comment, marked as answer]
> Is there really no unmountInactiveRoutes / unmountOnBlur for Stack Navigation

```js
const isFocused = useIsFocused();

if (!isFocused) {
  return null;
}

return <YourContent />;
```

> Is this conventional?

How many apps have you used where going back to a screen has all state reset? I haven't used any app that does this so it's not conventional and unexpected behaviour for the user.

**Comment by lcsjunior** [comment]
@satya164 Ok for useIsFocused. But what if, after an edit, on the previous screen I want a loading that awaits a fetch?

Simple example:

```js
  const [isLoading, setIsLoading] = useState(true);

  useFocusEffect(
    useCallback(() => {
      setIsLoading(true);
      (async () => {
        await sleep(3000);
        console.log('3 seconds later …');
        setIsLoading(false);
      })();
      return () => {
        console.log('Unfocused');
      };
    }, [])
  );

  if (isLoading) {
    return (
      <Center flex={1}>
        <Spinner size="lg" />
      </Center>
    );
  }
```
Is that what they do around there too?

We have an MVP where we usually need to reset the screen state back or the references based on the changed data in the nested screen.

**Reply by satya164** [reply]
Asynchronous data loading should be done by something like React Query or a similar approach where the inputs are dependencies of the query and outdated data can be easily invalidated. Resetting a screen isn't necessary in majority of cases - and is usually a sign of a bigger problem.

**Reply by lcsjunior** [reply]
I used SWR instead of React Query. But on a complex screen, I took this off to have more control of the solution.
In time, it looks like React Query has created a hook to solve this:

[https://tanstack.com/query/v4/docs/react-native](url)
![image](https://user-images.githubusercontent.com/8947621/200698217-55a71725-a331-4a50-abf6-d7739a3110c0.png)

There are hundreds of repos that use an alternate hook for this:

![image](https://user-images.githubusercontent.com/8947621/200699691-22edc833-cb39-4eaf-8991-b46fca72a852.png)

Most are codes that reference for listing: E.g.: Articles**List**Screen.js

![image](https://user-images.githubusercontent.com/8947621/200700137-eb3c0527-0590-48cb-80da-d4bc0956be23.png)

Others seem to override the back button:

[https://github.com/IronTony/react-native-react-query-starter-app/blob/main/src/hooks/useNavigationBack.ts](url)

**Comment by lcsjunior** [comment]
@satya164 you were correct. useIsFocused resolves.

Sample:
```js
const PanelListScreen = () => {
  const isFocused = useIsFocused();

  if (!isFocused) {
    return null;
  }

  return <PanelList />;
};
```

Thanks. Can you close this thread?

**Reply by satya164** [reply]
@lcsjunior you can mark my reply as the answer. it's just a discussion so there's no need to close
